src/net_processing.cpp

```diff
@@ -2643,10 +2642,17 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         // set nodes not capable of serving the complete blockchain history as "limited nodes"
         pfrom.m_limited_node = (!(nServices & NODE_NETWORK) && (nServices & NODE_NETWORK_LIMITED));
 
-        LOCK(peer->m_tx_relay_mutex);
-        if (peer->m_tx_relay != nullptr) {
+        // We don't initialize the m_tx_relay data structure if:
+        // - this is an outbound block-relay-only connection; or
+        // - fRelay=false and we're not offering NODE_BLOOM to this peer
+        //   (NODE_BLOOM means that the peer may turn on tx relay later)
```

Comment:
in 238ce0b02b1f7ac78c8c0466cef33628c6d49359:

If it is turned on later, we could keep it a nullptr until then and only initialize if and when we receive a `filter*` message?

Reply:
Perhaps, but I think it's better to allocate resources for the worst-case scenario up front, rather than try to reduce resource usage and later find out that you need to allocate a new structure.

Reply:
>rather than try to reduce resource usage and later find out that you need to allocate a new structure

What exactly is difficult about this?

Reply:
> What exactly is difficult about this?

I think I agree with John that there is little benefit (or even a risk) when giving remote peers the option to force you to allocate MBs of memory by sending a single msg, instead of allocating the memory on startup/creation of the connection.

Reply:
Hmm, allocating upfront is like giving a remote peer the option and he always exercises it immediately after connecting, no?

Reply:
Yes, and I think this is preferable because there are less (late) surprises. Neither on the live network, nor when developing. For example, `massif` will display the real usage from the start. Or, imagine someone under-estimating the resource usage and thus increasing the maxconnections count, then running into OOM later on.

Reply:
The counter scenario is somebody running into OOM because MBs of memory were allocated needlessly (allocated but never used because the peer(s) never sent `filter*`).

Reply:
I think if you want to reduce memory usage, you should disable bip37, not enable it with the hope that no one (or only a small fraction of your incoming peers) will use it.